agent/uiserver/uiserver.go

```diff
@@ -0,0 +1,190 @@
+	buf, info, err := renderIndex(newCfg, fs)
+	if _, ok := err.(*os.PathError); ok && newCfg.UIConfig.Dir != "" {
+		// A Path error indicates that there is no index.html. This could happen if
+		// the user configured their own UI dir and is serving something that is not
+		// our usual UI. This won't work perfectly because our uiserver will still
+		// redirect everything to the UI but we shouldn't fail the entire UI server
+		// with a 500 in this case. Partly that's just bad UX and partly it's a
+		// breaking change although quite an edge case. Instead, continue but just
+		// return a 404 response for the index.html and log a warning.
+		h.logger.Warn("ui_config.dir does not contain an index.html. Index templating and redirects to index.html are disabled.")
+	} else if err != nil {
+		return err
+	}
+
+	// buf can be nil in the PathError case above. We should skip this part but
+	// still serve the rest of the files in that case.
+	if buf != nil {
+		// Create a new fs that serves the rendered index file or falls back to the
+		// underlying FS.
+		fs = &bufIndexFS{
+			fs:            fs,
+			indexRendered: buf,
+			indexInfo:     info,
+		}
+
+		// Wrap the buffering FS our redirect FS. This needs to happen later so that
+		// redirected requests for /index.html get served the rendered version not the
+		// original.
+		fs = &redirectFS{fs: fs}
+	}
+
+	newState.srv = http.FileServer(fs)
+
+	// Store the new state
+	h.state.Store(newState)
+	return nil
+}
+
+// getState is a helper to access the atomic internal state
+func (h *Handler) getState() *reloadableState {
+	if cfg, ok := h.state.Load().(reloadableState); ok {
+		return &cfg
+	}
+	return nil
+}
+
+func renderIndex(cfg *config.RuntimeConfig, fs http.FileSystem) ([]byte, os.FileInfo, error) {
+	// Open the original index.html
+	f, err := fs.Open("/index.html")
+	if err != nil {
+		return nil, nil, err
+	}
+	defer f.Close()
+
+	content, err := ioutil.ReadAll(f)
+	if err != nil {
+		return nil, nil, fmt.Errorf("failed reading index.html: %s", err)
+	}
+	info, err := f.Stat()
+	if err != nil {
+		return nil, nil, fmt.Errorf("failed reading metadata for index.html: %s", err)
+	}
+
+	// Create template data from the current config.
+	tplData, err := uiTemplateDataFromConfig(cfg)
+	if err != nil {
+		return nil, nil, fmt.Errorf("failed loading UI config for template: %s", err)
+	}
+
+	// Sadly we can't perform all the replacements we need with Go template
+	// because some of them end up being rendered into an escaped json encoded
+	// meta tag by Ember build which messes up the Go template tags. After a few
+	// iterations of grossness, this seemed like the least bad for now. note we
+	// have to match the encoded double quotes around the JSON string value that
+	// is there as a placeholder so the end result is an actual JSON bool not a
+	// string containing "false" etc.
+	re := regexp.MustCompile(`%22__RUNTIME_BOOL_[A-Za-z0-9-_]+__%22`)
```

Comment:
This could be a package constant.

Reply:
Since it's not on a hot path (typically called once per agent lifetime) and very unlikely to be useful in any other place do you think it's preferable to define it out of context on the package? I don't mind but my feeling was it's easier to read this code with it inline so you can see what it's doing all at once.